service/history/replication/ack_manager.go

```diff
@@ -256,53 +258,64 @@ func (p *ackMgrImpl) getTasks(
 	maxTaskID int64,
 	batchSize int,
 ) ([]*replicationspb.ReplicationTask, int64, error) {
-
 	if minTaskID > maxTaskID {
 		return nil, 0, serviceerror.NewUnavailable("min task ID < max task ID, probably due to shard re-balancing")
 	} else if minTaskID == maxTaskID {
 		return []*replicationspb.ReplicationTask{}, maxTaskID, nil
 	}
 
-	var token []byte
 	replicationTasks := make([]*replicationspb.ReplicationTask, 0, batchSize)
-	for {
-		response, err := p.executionMgr.GetHistoryTasks(ctx, &persistence.GetHistoryTasksRequest{
-			ShardID:             p.shard.GetShardID(),
-			TaskCategory:        tasks.CategoryReplication,
-			InclusiveMinTaskKey: tasks.NewImmediateKey(minTaskID + 1),
-			ExclusiveMaxTaskKey: tasks.NewImmediateKey(maxTaskID + 1),
-			BatchSize:           batchSize,
-			NextPageToken:       token,
-		})
+	iter := collection.NewPagingIterator(p.getPaginationFn(ctx, minTaskID, maxTaskID, batchSize))
+	for iter.HasNext() && len(replicationTasks) < batchSize {
```

Comment:
why need to check len(replicationTasks) < batchSize ?

Reply:
if the min ID -> max ID contains many tasks?

Reply:
I am swapping these checks to avoid extra read in case if `replicationTasks` is already full (#3641).